packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,478 @@
+	const startRef = string.createLocalReferencePosition(
+		startSeg,
+		interval.start.getOffset(),
+		ReferenceType.StayOnRemove,
+		startSeg.properties,
+	);
+	const endRef = string.createLocalReferencePosition(
+		endSeg,
+		interval.end.getOffset(),
+		ReferenceType.StayOnRemove,
+		endSeg.properties,
+	);
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: startRef,
+		end: endRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: SharedStringRevertible[],
+) {
+	// This should not be called if the interval slid because of a string remove
+
+	const startSeg = previousInterval.start.getSegment() as SharedStringSegment;
+	const endSeg = previousInterval.end.getSegment() as SharedStringSegment;
+	const prevStartRef = string.createLocalReferencePosition(
+		startSeg,
+		previousInterval.start.getOffset(),
+		ReferenceType.StayOnRemove,
+		startSeg.properties,
+	);
+	const prevEndRef = string.createLocalReferencePosition(
+		endSeg,
+		previousInterval.end.getOffset(),
+		ReferenceType.StayOnRemove,
+		endSeg.properties,
+	);
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: prevStartRef,
+		end: prevEndRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: SharedStringRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+
+function addIfIntervalEndpoint(
+	ref: LocalReferencePosition,
+	segmentIndex: number,
+	startIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+	endIntervals: { segmentIndex: number; interval: SequenceInterval }[],
+) {
+	if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeBegin)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			startIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	} else if (refTypeIncludesFlag(ref.refType, ReferenceType.RangeEnd)) {
+		const interval = ref.properties?.interval;
+		if (interval && interval instanceof SequenceInterval) {
+			endIntervals.push({ segmentIndex, interval });
+			return true;
+		}
+	}
+	return false;
+}
+
+function addLocalRefPlaceholders(string: SharedString, ref: LocalReferencePosition) {
+	return {
+		stay: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.StayOnRemove,
+			undefined,
+		),
+		slide: string.createLocalReferencePosition(
+			ref.getSegment() as SharedStringSegment,
+			ref.getOffset(),
+			ReferenceType.SlideOnRemove,
+			undefined,
+		),
+	};
+}
+
+export function appendToMergeTreeDeltaRevertibles(
+	string: SharedString,
```

Comment:
does typescript let you use the reserved word string as a param/var name? seems odd, maybed call it sharedString

Reply:
It does, but I agree it looks odd.